### PR TITLE
Filecount plugin: Custom values for reported plugin, type and type instance

### PR DESCRIPTION
**Pull request by rpv-tomsk**
Hi!

Proposed patch adds ability to make flexible configuration of `filecount` plugin:
- Allow to set plugin name;
- Allow to set empty plugin instance;
- Allow to set metrics type: 'FilesSizeType' and 'FilesCountType' options added;
- Allow to disable metric reporting;
- Allow to set metrics type instance ('TypeInstance' option added).

Continues work of #1944.

Configuration example:

```
<Plugin filecount>
 <Directory "/var/spool/postfix/active">
    Plugin "postfix"
    Instance ""
    TypeInstance "active"
 </Directory>
 <Directory "/var/spool/postfix/deferred">
    Plugin "postfix"
    Instance ""
    TypeInstance "deferred"
 </Directory>
 <Directory "/var/spool/postfix/incoming">
    Plugin "postfix"
    Instance ""
    TypeInstance "incoming"
 </Directory>
 <Directory "/var/spool/postfix/maildrop">
    Plugin "postfix"
    Instance ""
    TypeInstance "maildrop"
 </Directory>
 <Directory "/var/spool/postfix/hold">
    Plugin "postfix"
    Instance ""
    TypeInstance "hold"
 </Directory>
</Plugin>
```

### CODE REVIEW

File: src/collectd.conf.in [hunk 1]
```diff
@@ -558,12 +558,42 @@
...
 #		Name "*.conf"
 #		MTime "-5m"
 #		Size "+10k"
 #		Recursive true
 #		IncludeHidden false
+#		#FilesSizeType "bytes"
+#		#FilesCountType "files"
+#		#TypeInstance "instance"
+#	</Directory>
+#	#Postfix mail queue size
```
Comment (octo) [inline src/collectd.conf.in]: Such example configurations should either go in the `contrib/` directory or (preferably) in the wiki. Some plugins already have a "…/Config" subpage and the [FileCount plugin](https://collectd.org/wiki/index.php/Plugin:FileCount) could have one, too.

File: src/collectd.conf.pod [hunk 2]
```diff
@@ -2585,12 +2585,16 @@ blocks, the following options are recognized:
 
 =over 4
 
+=item B<Plugin> I<Plugin>
+
+Use I<Plugin> as the plugin name when submitting values.
+Defaults to 'filecount'.
```
Comment (octo) [inline src/collectd.conf.pod]: Please use `B<…>` for literal values, i..e `B<filecount>`.

File: src/collectd.conf.pod [hunk 3]
```diff
@@ -2636,6 +2640,21 @@ Controls whether or not to include "hidden" files and directories in the count.
 "Hidden" files and directories are those, whose name begins with a dot.
 Defaults to I<false>, i.e. by default hidden files and directories are ignored.
 
+=item B<FilesSizeType> I<Type>
+
+Sets the type used to dispatch files combined size. Empty value ("") disables
+reporting. Defaults to 'bytes'.
```
Comment (octo) [inline src/collectd.conf.pod]: Dito: `B<bytes>`

File: src/collectd.conf.pod [hunk 4]
```diff
@@ -2636,6 +2640,21 @@ Controls whether or not to include "hidden" files and directories in the count.
...
+
+=item B<FilesCountType> I<Type>
+
+Sets the type used to dispatch number of files. Empty value ("") disables
+reporting. Defaults to 'files'.
```
Comment (octo) [inline src/collectd.conf.pod]: Dito: `B<files>`

**Comment by rpv-tomsk** [comment]
Patch updated.

File: src/collectd.conf.in (hunk 1 above)
Comment (rpv-tomsk) [inline src/collectd.conf.in]: Where should we place note to update Wiki page after this will be released?

**Comment by rpv-tomsk** [comment]
Patch updated: Do not set vl.plugin_instance to ''.

Amended:
```
--- a/src/filecount.c
+++ b/src/filecount.c
@@ -72,7 +72,8 @@ static void fc_submit_dir (const fc_directory_conf_t *dir)
   sstrncpy (vl.plugin,
             (dir->plugin_name != NULL) ? dir->plugin_name : "filecount",
             sizeof (vl.plugin));
-  sstrncpy (vl.plugin_instance, dir->instance, sizeof (vl.plugin_instance));
+  if (dir->instance != NULL)
+    sstrncpy (vl.plugin_instance, dir->instance, sizeof (vl.plugin_instance));
   if (dir->type_instance != NULL)
     sstrncpy(vl.type_instance, dir->type_instance, sizeof(vl.type_instance));

@@ -428,6 +429,9 @@ static int fc_config_add_dir (oconfig_item_t *ci)
     status = -1;
   }

+  if (strlen(dir->instance) == 0)
+    sfree(dir->instance);
+
   if (status == 0)
   {
     fc_directory_conf_t **temp;
```

**Comment by rpv-tomsk** [comment]
ping

**Comment by rpv-tomsk** [comment]
Ping me if you want this PR to be merged, I will rebase it to current master.

**Comment by rpv-tomsk** [comment]
The PR rebased to current master.

Please take a look and merge it.

**Comment by rpv-tomsk** [comment]
Rebased again, added some code style cleanup.

**Pull request merged**